C++ build fails when configured with RDKIT_USE_BOOST_SERIALIZATION=OFF

<!-- Please help us help you: We normally need the information below in order to answer questions or understand bug reports. If you do not provide the information requested we may not be able to help you and will probably close the issue. --> 

**Configuration:**

- RDKit Version: 2020.09.1 and current tip of the master branch
- Operating system: observed on linux, it could probably affect other OSs too.

**Description:**
When the cmake build is configured with the `RDKIT_USE_BOOST_SERIALIZATION` flag disabled, the C++ build stops with an error inside `Code/GraphMol/ScaffoldNetwork` that starts with the output below (some conditional compilation directives are probably missing inside `ScaffoldNetwork.h`).

```
Scanning dependencies of target ScaffoldNetwork
[ 99%] Building CXX object Code/GraphMol/ScaffoldNetwork/CMakeFiles/ScaffoldNetwork.dir/ScaffoldNetwork.cpp.o
In file included from /opt/RDKit-build/rdkit/Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.cpp:10:
/opt/RDKit-build/rdkit/Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.h:188:8: error: 'version' is not a class template
  188 | struct version<RDKit::ScaffoldNetwork::ScaffoldNetwork> {
      |        ^~~~~~~
/opt/RDKit-build/rdkit/Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.h:188:57: error: explicit specialization of non-template 'boost::serialization::version'
  188 | struct version<RDKit::ScaffoldNetwork::ScaffoldNetwork> {
      |                                                         ^
In file included from /usr/include/boost/multi_index/detail/hash_index_iterator.hpp:22,
                 from /usr/include/boost/multi_index/hashed_index.hpp:33,
                 from /usr/include/boost/graph/named_graph.hpp:18,
                 from /usr/include/boost/graph/adjacency_list.hpp:37,
                 from /opt/RDKit-build/rdkit/Code/GraphMol/ROMol.h:27,
                 from /opt/RDKit-build/rdkit/Code/GraphMol/GraphMol.h:19,
                 from /opt/RDKit-build/rdkit/Code/GraphMol/RDKitBase.h:26,
                 from /opt/RDKit-build/rdkit/Code/GraphMol/ScaffoldNetwork/detail.h:12,
                 from /opt/RDKit-build/rdkit/Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.cpp:11:
/usr/include/boost/serialization/version.hpp:36:8: error: 'boost::serialization::version' is not a template
```

<!-- If you have a support question, please use the rdkit-discuss mailing list: 
https://sourceforge.net/p/rdkit/mailman/ -->

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -89,10 +89,10 @@ if(RDK_BUILD_SWIG_WRAPPERS!=ON)
 endif()
 
 if(RDK_SWIG_STATIC AND RDK_BUILD_SWIG_WRAPPERS)
-  if(NOT RDK_INSTALL_STATIC_LIBS)
+  if(NOT MSVC AND NOT RDK_INSTALL_STATIC_LIBS)
     message("Enabling RDK_INSTALL_STATIC_LIBS because RDK_SWIG_STATIC is set.")
     set(RDK_INSTALL_STATIC_LIBS ON CACHE BOOL "install the rdkit static libraries" FORCE)
-  endif(NOT RDK_INSTALL_STATIC_LIBS)
+  endif(NOT MSVC AND NOT RDK_INSTALL_STATIC_LIBS)
 endif()
 
 if(NOT (MSVC AND (NOT RDK_INSTALL_DLLS_MSVC)))
@@ -100,10 +100,10 @@ if(NOT (MSVC AND (NOT RDK_INSTALL_DLLS_MSVC)))
 endif()
 
 if(RDK_PGSQL_STATIC AND RDK_BUILD_PGSQL)
-  if(NOT RDK_INSTALL_STATIC_LIBS)
+  if(NOT MSVC AND NOT RDK_INSTALL_STATIC_LIBS)
     message("Enabling RDK_INSTALL_STATIC_LIBS because RDK_PGSQL_STATIC is set.")
     set(RDK_INSTALL_STATIC_LIBS ON CACHE BOOL "install the rdkit static libraries" FORCE)
-  endif(NOT RDK_INSTALL_STATIC_LIBS)
+  endif(NOT MSVC AND NOT RDK_INSTALL_STATIC_LIBS)
 endif()
 
 
@@ -390,7 +390,10 @@ if(RDK_USE_BOOST_SERIALIZATION)
     else()
       set(Boost_LIBRARIES ${T_LIBS})
     endif()
-  target_compile_definitions(rdkit_base INTERFACE -DRDK_USE_BOOST_SERIALIZATION -DBOOST_SERIALIZATION_DYN_LINK)
+    target_compile_definitions(rdkit_base INTERFACE -DRDK_USE_BOOST_SERIALIZATION)
+    if(NOT Boost_USE_STATIC_LIBS)
+      target_compile_definitions(rdkit_base INTERFACE -DBOOST_SERIALIZATION_DYN_LINK)
+    endif()
 endif()
 
 

--- a/Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.h
+++ b/Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.h
@@ -182,6 +182,7 @@ RDKIT_SCAFFOLDNETWORK_EXPORT ScaffoldNetworkParams getBRICSNetworkParams();
 }  // namespace ScaffoldNetwork
 }  // namespace RDKit
 
+#ifdef RDK_USE_BOOST_SERIALIZATION
 namespace boost {
 namespace serialization {
 template <>
@@ -190,5 +191,5 @@ struct version<RDKit::ScaffoldNetwork::ScaffoldNetwork> {
 };
 }  // namespace serialization
 }  // namespace boost
-
+#endif
 #endif

